llama_index/callbacks/base.py

```diff
@@ -88,3 +113,36 @@ def remove_handler(self, handler: BaseCallbackHandler) -> None:
     def set_handlers(self, handlers: List[BaseCallbackHandler]) -> None:
         """Set handlers as the only handlers on the callback manager."""
         self.handlers = handlers
+
+    @contextmanager
+    def as_trace(self, run_id: str) -> Generator[None, None, None]:
+        """Context manager tracer for lanching and shutdown of runs."""
+        self.launch(run_id=run_id)
+        yield
+        self.shutdown(run_id=run_id)
+
+    def launch(self, run_id: Optional[str] = None) -> None:
+        """Run when an overall run is launched."""
+        if not self._running_id:
+            self._trace_map = defaultdict(list)
+            self._trace_stack = [BASE_TRACE_ID]
+
+            for handler in self.handlers:
+                handler.launch(run_id=run_id)
+
+            self._running_id = run_id

```

Comment:
should we error if _running_id is already specified?

Reply:
Actually no! Since this is just used to track the entry point of the trace (index construction, query, insert, etc.)

A single entry point might call launch for other entry points, but we don't want to call launch on all the handlers again